sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -2349,7 +2350,22 @@ class Analyzer(
   }
 
   /**
-   * Replace [[TimeZoneAwareExpression]] without timezone id by its copy with session local
+   * Replace unresolved expressions in grouping keys with resolved ones in SELECT clauses.
+   */
+  object ResolveAggAliasInGroupBy extends Rule[LogicalPlan] {
+
+    override def apply(plan: LogicalPlan): LogicalPlan = plan.resolveOperators {
+      case agg @ Aggregate(groups, aggs, child)
+          if conf.groupByAliases && child.resolved && groups.exists(!_.resolved) =>
+        agg.copy(groupingExpressions = groups.map {
+          case u: UnresolvedAttribute => aggs.find(ne => resolver(ne.name, u.name)).getOrElse(u)
```

Comment:
In MySQL, it behaves quite similarly:

mysql> select count(a) as c from test group by c;
ERROR 1056 (42000): Can't group on 'c'

I think an thrown exception is good.